packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1250,7 +1234,8 @@ function dispatchAction<S, A>(
       const eagerReducer = queue.eagerReducer;
       if (eagerReducer !== null) {
         if (__DEV__) {
-          ReactCurrentDispatcher.current = InvalidHooksDispatcherOnUpdateInDEV;
+          hooksAccessForbiddenInDEV = true;
```

Comment:
One of the motivations for this refactor in the first place was to get rid of additional flags like this that is difficult to keep track of whether they're all assigned together in the same place and how the relate to assignments of the dispatcher. By adding this in again, I've now lost confidence that I understand these pairs again.

Can we make it just part of the dispatcher again?

Reply:
I went with that approach originally but it became a problem after I made the change to switch back to the mount dispatcher once we reach the end of the list

https://github.com/facebook/react/blob/5a739725cd126e9cb1353279c64d46617712b3b7/packages/react-reconciler/src/ReactFiberHooks.js#L567-L572

because now the `finally` block will reset the dispatcher back to the old one.

An alternative would be to wrap each call site with a separate `try/finally` block but that leads to lots of forking in the code because you have to store a reference to the previous dispatcher. And it still suffers from the same problem because the nested hook could mutate the dispatcher — which is perhaps fine because that's a buggy code path, anyway, but it's a deviation from production behavior.

`hooksAccessForbiddenInDEV` seemed like the best trade off, and I don't think it's as bad as the previous boolean because it's completely localized within the DEV block that defines the DEV dispatcher.

Reply:
I don't like that we're adding the progressive enhancement constraint before we need it. Isn't the simple fix here to just disable that feature and not reset it?

It seems like a problematic pattern in general if we can't use try/finally in that way since we're now deviating from the "algebraic effects" approach.

Reply:
> It seems like a problematic pattern in general if we can't use try/finally in that way since we're now deviating from the "algebraic effects" approach.

I don't think is a problem in general, it's only a problem because of the unsupported nested mutation, which is always a bug and only exists because we don't want to add a redundant check to the prod version.

Reply:
The other thing I considered was putting `createWorkInProgress` on the Dispatcher.

Reply:
Once we have released the first version, we have plenty of time to figure out how we could structure the progressive enhancement case in a clever optimal way.

Reply:
For real progressive enhancement we probably need a way to validate that it is a legit usage of expanding properties, e.g. by inserting a special primitive that when called creates a new tree. But we might also need the nested case. So I'm not sure this solution works anyway.

Reply:
Resolved via chat: We don't support progressive enhancement yet anyway so let's throw and follow up once we get to that feature.

Reply:
@acdlite Could I ask what is "progressive enhancement" in this context?